src/types/recovery.rs

```diff
@@ -0,0 +1,201 @@
+    /// R value.
+    pub r: H256,
+    /// S value.
+    pub s: H256,
+}
+
+impl Recovery {
+    /// Creates new recovery data from its parts.
+    pub fn new<M>(message: M, v: u64, r: H256, s: H256) -> Recovery
+    where
+        M: Into<RecoveryMessage>,
+    {
+        Recovery {
+            message: message.into(),
+            v,
+            r,
+            s,
+        }
+    }
+
+    /// Creates new recovery data from a raw signature.
+    ///
+    /// This parses a raw signature which is expected to be 65 bytes long where
+    /// the first 32 bytes is the `r` value, the second 32 bytes the `s` value
+    /// and the final byte is the `v` value in 'Electrum' notation.
+    pub fn from_raw_signature<M, B>(message: M, raw_signature: B) -> Result<Recovery, ParseSignatureError>
+    where
+        M: Into<RecoveryMessage>,
+        B: AsRef<[u8]>,
+    {
+        let bytes = raw_signature.as_ref();
+
+        if bytes.len() != 65 {
+            return Err(ParseSignatureError);
+        }
+
+        let v = bytes[64];
+        let r = H256::from_slice(&bytes[0..32]);
+        let s = H256::from_slice(&bytes[32..64]);
+
+        Ok(Recovery::new(message, v as _, r, s))
+    }
+
+    /// Retrieves the recovery signature in standard notation.
+    ///
+    /// This method returns a `ethsign::Signature` with `v` in standard
+    /// notation. This means that the `27`, `28` or chain relay protection is
+    /// removed from the recovery's v value.
+    pub fn as_signature(&self) -> Signature {
+        Signature {
+            v: self.standard_v(),
+            r: self.r.into(),
+            s: self.s.into(),
+        }
+    }
+
+    /// Retrieve recovery value `v` in standard notation.
+    pub fn standard_v(&self) -> u8 {
+        // this is a re-implementation of `ethereum-transaction` as there is no
+        // good way to call this without building a full `SignedTransaction`
+        // which isn't needed here
+        match self.v {
+            27 => 0,
+            28 => 1,
+            v if v >= 35 => ((v - 1) % 2) as _,
+            _ => 4,
+        }
+    }
+}
+
+impl<'a> From<&'a SignedData> for Recovery {
+    fn from(signed: &'a SignedData) -> Self {
+        Recovery::new(signed.message_hash, signed.v as _, signed.r, signed.s)
+    }
+}
+
+impl<'a> From<&'a SignedTransaction> for Recovery {
+    fn from(tx: &'a SignedTransaction) -> Self {
+        Recovery::new(tx.message_hash, tx.v, tx.r, tx.s)
+    }
+}
+
+impl<'a> From<&'a EthtxSignedTransaction<'a>> for Recovery {
+    fn from(tx: &'a EthtxSignedTransaction<'a>) -> Self {
+        Recovery::new(tx.bare_hash(), tx.v, H256(tx.r.into()), H256(tx.s.into()))
+    }
+}
+
+/// Recovery message data.
+///
+/// The message data can either be a string message that is first hashed
+/// according to EIP-191 and then recovered based on the signature or a
+/// precomputed hash.
+#[derive(Clone, Debug, PartialEq)]
+pub enum RecoveryMessage {
+    /// Message string
+    String(String),
```

Comment:
Shouldn't that be `Data(Bytes)`? I think you can sign arbitraty (non-utf8) bytes.

Reply:
I see that we have restricted hashing to `AsRef<str>` anyway, so it's good as is.

Reply:
Wait a second, that's a good point - shouldn't this crate allow hashing arbitrary data?

Reply:
Yeah, feel free to do a follow up PR on this, or maybe @nlordell is interested :)

Reply:
Looks like @reuvenpo beat me to it!